wetterdienst/ui/app.py

```diff
@@ -0,0 +1,225 @@
+app.layout = dashboard_layout()
+
+OBSERVATION_VALUES_PARAMETER_COLUMN = "parameter"
+OBSERVATION_VALUES_VALUE_COLUMN = "value"
+OBSERVATION_VALUES_DATE_COLUMN = "date"
+
+
+@app.callback(
+    Output("hidden-div-metadata", "children"),
+    [
+        Input("select-parameter", "value"),
+        Input("select-time-resolution", "value"),
+        Input("select-period-type", "value"),
+    ],
+)
+def update_meta_data(parameter, time_resolution, period_type):
+    """
+    Function to update the metadata according to
+    the selection of the dropdowns
+    It stores MetaData behind a hidden div on the front-end
+    """
+    log.info(
+        f"Requesting stations for "
+        f"parameter={parameter}, "
+        f"resolution={time_resolution}, "
+        f"period={period_type}"
+    )
+    try:
+        stations = DwdObservationRequest(
+            parameter=DwdObservationDataset(parameter),
+            resolution=DwdObservationResolution(time_resolution),
+            period=DwdObservationPeriod(period_type),
+        ).all()
+    except InvalidParameterCombination:
+        raise PreventUpdate
+
+    df = stations.df
+
+    log.info(f"Forwarding stations data frame {frame_summary(df)}")
+
+    return df.to_json(date_format="iso", orient="split")
+
+
+@app.callback(
+    Output("graph1", "figure"),
+    [Input("select-variable", "value")],
+    [State("hidden-div", "children")],
+)
+def make_graph(variable, jsonified_data):
+    """  takes hidden data to show up the central plot  """
+
+    # FIXME: This does not work yet.
+    if not jsonified_data:
+        return html.Div(html.P("No data available"))
+
+    climate_data = pd.read_json(jsonified_data, orient="split")
+    log.info(
+        f"Building graph for variable={variable} from {frame_summary(climate_data)}"
+    )
+
+    fig = default_figure(climate_data, variable)
+    fig.update_layout(
+        margin=go.layout.Margin(
+            l=0,  # left margin
+            r=0,  # right margin
+            b=0,  # bottom margin
+            t=0,  # top margin
+        )
+    )
+    return fig
+
+
+@app.callback(
+    Output("hidden-div", "children"),
+    [Input("select-weather-stations", "value")],
+    [
+        State("select-parameter", "value"),
+        State("select-time-resolution", "value"),
+        State("select-period-type", "value"),
+    ],
+)
+def update_data(
+    station_id: int, parameter: str, time_resolution: str, period_type: str
+):
+    """ stores selected data behind a hidden div box to share with other callbacks """
+    log.info(
+        f"Requesting values for "
+        f"station_id={station_id}, "
+        f"parameter={parameter}, "
+        f"resolution={time_resolution}, "
+        f"period={period_type}"
+    )
+    stations = DwdObservationRequest(
+        parameter=DwdObservationDataset(parameter),
+        resolution=DwdObservationResolution(time_resolution),
+        period=DwdObservationPeriod(period_type),
+        tidy=False,
```

Comment:
I guess we should try and go via tidy here and use .groupby to create the json response. Maybe we can also create another output format for ValuesResult, that is meant to be used for the ui only?

Reply:
I believe, for the first iteration, it will be fine. Pandas data frames in non-tidy format work just perfectly here. Another variant should be added carefully, maybe within another iteration.